### PR TITLE
WSTEAM1-725 CPS SPIKE:Live page archive support

### PR DESCRIPTION
**Pull request by emilysaffron**
Resolves JIRA [WSTEAM1-725](https://jira.dev.bbc.co.uk/browse/WSTEAM1-725)

Overall changes
======
Routes CPS live pages to be rendered by Simorgh using a query param override i.e the `renderer_env`.

Code changes
======
- creates `arabic` live page data fixture
- Seperates logic for `LIVE` and `TOPIC` pages inside of `constructPageFetchUrl` function (previously they were grouped together), and refactors existing logic for when the environment is local.
- Updates existing test cases.
- Adds integration tests.
- Updates existing snapshots.

Testing
======
1. Pull down PR
2. go to http://localhost:7081/arabic/live/67574192?renderer_env=live and see it is [this cps live page ](https://www.test.bbc.com/arabic/live/67574192) being rendered by simorgh
3. Similarly go to http://localhost:7081/serbian/lat/live/media-23179005 which has been added as a local fixture using https://www.test.bbc.com/serbian/lat/live/media-23179005?morph_env=test

Helpful Links
======
_Add Links to useful resources related to this PR if applicable._

[Gap analysis between Simorgh and CPS render](https://paper.dropbox.com/doc/Spike-Live-Page-gap-analysis-Simorgh-vs-CPS-Live--CGCg5Marbmh0IYCImBA3JHJnAg-e162RpJOpwUnNJF55qzws)

[Coding Standards](https://github.com/bbc/simorgh/blob/latest/docs/Coding-Standards/README.md)

[Repository use guidelines](https://github.com/bbc/simorgh-infrastructure/blob/latest/documentation/repository-guidelines.md)

### CODE REVIEW

File: src/app/routes/utils/constructPageFetchUrl/index.ts
Comment (amoore108) [inline src/app/routes/utils/constructPageFetchUrl/index.ts:99]: I think you can maybe get rid of passing around the `isArchive` flag and do the URL checking under the `case LIVE_PAGE` case, maybe something like:

```
getIdFunction = (path: string) => {
        if (!path.match(/(c[a-zA-Z0-9]{10}t)/)) {
          const cpsId = getCpsId(path);
          return `/${service}/live/${cpsId}`;
        }
        return getTipoId(path);
      };
```

That would check that if the ID is _not_ a TIPO ID structure, then assume CPS, or else return the TIPO ID.

It doesn't add the `renderer_env` flag though, you'd need to manually specify `renderer_env=live` in the URL and update the `[[...variant]].page.tsx` file to allow the `rendererEnv` prop to be set, as currently its defaulting always to `test`.

Comment (emilysaffron) [reply src/app/routes/utils/constructPageFetchUrl/index.ts:99]: oh wow yea this is much simpler thank you! will implement now 🥳

File: src/app/routes/utils/constructPageFetchUrl/index.ts
Comment (LilyL0u) [inline src/app/routes/utils/constructPageFetchUrl/index.ts:99]: Note from huddle: do we want this to have its own path matching code here? /service/live/serviceId should move to the LIVE_PAGE case. Then we don't blur the lines between the different page types.

File: ws-nextjs-app/pages/[service]/live/[id]/[[...variant]].page.tsx
Comment (LilyL0u) [inline ws-nextjs-app/pages/[service]/live/[id]/[[...variant]].page.tsx]: Whatever the live page code is it will always default the live pages to fetch from test fabl, but we want to control this with the rendere_env instead. No hard coding allowed 👿

File: src/app/routes/utils/constructPageFetchUrl/index.test.ts
Comment (LilyL0u) [inline src/app/routes/utils/constructPageFetchUrl/index.test.ts]: We also want integration tests for this.  Makes sure the local routing works.  Grab the data for the arabic live page, put into fixture data, assert that we can render the arabic live page. Then next pieces of work we can decide whether to add more cypress tests on the next js work.

File: src/app/routes/utils/constructPageFetchUrl/index.test.ts
Comment (LilyL0u) [inline src/app/routes/utils/constructPageFetchUrl/index.test.ts:106]: I have changed the pathnames to only include the id which is just the numbers and not the /arabic/live/ substring, because in the real code the pathname is passed into constructPageFetchUrl as just the numbers. We want the rest of the substring added _within_  the constructPageFetchUrl function for the bff data fetch url, but I haven't yet traced down the reason for the numbers part of the id being passed into getPageData alone. Perhaps the numbers part of the id is used alone elsewhere (e.g routing).

File: src/app/routes/utils/constructPageFetchUrl/index.ts
Comment (LilyL0u) [inline src/app/routes/utils/constructPageFetchUrl/index.ts]: Shorter alternative to achieve stripping /${service/live from the id:

```
const newId = id.replace(/^\/?[^/]+\/live\//, '');
        
        fetchUrl = Url(
          `${host}${port}/api/local/${service}/live/${newId}${variantPath}`,
        );
```

File: src/app/routes/utils/constructPageFetchUrl/index.ts
Comment (LilyL0u) [inline src/app/routes/utils/constructPageFetchUrl/index.ts:107]: would this return null even without explicitly returning null if none of the if conditions were satisfied?

File: src/app/routes/utils/constructPageFetchUrl/index.ts
Comment (karinathomasbbc) [inline src/app/routes/utils/constructPageFetchUrl/index.ts]: This can be simplified
```suggestion
          return  id;
```

File: src/app/routes/utils/constructPageFetchUrl/index.ts
Comment (karinathomasbbc) [inline src/app/routes/utils/constructPageFetchUrl/index.ts]: Do we need to know if the ID is a CPS id? Could we assume that we know we're rendering a live page, and therefore if it's not a Tipo ID, then it must be a CPS one?

If so, it means that this logic could be greatly simplified? 

Also, it would mean that there would be no need to "reconstruct" the ID (to be /service/live/id), because it is already the same as the path 

e.g. in pseudo code:
```
if isTipoId
 return id
else 
 return path
```

File: src/app/routes/utils/constructPageFetchUrl/index.ts
Comment (karinathomasbbc) [inline src/app/routes/utils/constructPageFetchUrl/index.ts]: No, the ID validation already happens in Belfrage, so no need for it here
We know that we're rendering a live page because of the existence of `/live/` in the URL

File: src/app/routes/utils/constructPageFetchUrl/index.ts
Comment (karinathomasbbc) [inline src/app/routes/utils/constructPageFetchUrl/index.ts]: @LilyL0u , @louisearchibald and I have had a discussion offline about how this could be simplified, using the `path` instead of `id` + regexes

File: src/app/routes/utils/constructPageFetchUrl/index.ts
Comment (LilyL0u) [reply src/app/routes/utils/constructPageFetchUrl/index.ts]: We decided to keep the CPS id check because we need to return null in the LIVE_PAGE case for the test that checks an error message for when the ID is not a valid tipo or cps one.

File: src/app/routes/utils/constructPageFetchUrl/index.ts
Comment (louisearchibald) [reply src/app/routes/utils/constructPageFetchUrl/index.ts]: Thanks for clarification, have removed the comment :)

File: ws-nextjs-app/integration/pages/live/arabic/amp.test.ts
Comment (karinathomasbbc) [inline ws-nextjs-app/integration/pages/live/arabic/amp.test.ts]: AMP is not supported for Live Pages, therefore this can be deleted

File: src/app/routes/utils/constructPageFetchUrl/index.ts
Comment (karinathomasbbc) [reply src/app/routes/utils/constructPageFetchUrl/index.ts]: OK that's fair enough. Technically routing in Belfrage would offer us this extra protection, but there's no harm in keeping it for now.

File: src/app/routes/utils/constructPageFetchUrl/index.ts
Comment (karinathomasbbc) [inline src/app/routes/utils/constructPageFetchUrl/index.ts:230]: Could we add a comment here to explain?

Something along the lines of: pathname is the ID of the live page, without `/service/live/`, and supports both Tipo & CPS IDs?

File: src/app/routes/utils/constructPageFetchUrl/index.ts
Comment (eagerterrier) [inline src/app/routes/utils/constructPageFetchUrl/index.ts]: remove?

File: src/app/routes/utils/constructPageFetchUrl/index.ts
Comment (eagerterrier) [inline src/app/routes/utils/constructPageFetchUrl/index.ts]: remove?

File: src/app/routes/utils/constructPageFetchUrl/index.ts
Comment (eagerterrier) [inline src/app/routes/utils/constructPageFetchUrl/index.ts]: remove?

File: ws-nextjs-app/pages/[service]/live/[id]/[[...variant]].page.tsx
Comment (eagerterrier) [inline ws-nextjs-app/pages/[service]/live/[id]/[[...variant]].page.tsx]: remove?

File: src/app/routes/utils/constructPageFetchUrl/index.ts
Comment (eagerterrier) [inline src/app/routes/utils/constructPageFetchUrl/index.ts]: think we can remove this line to make it easier to read / remove one conditional.

e.g. 



```
if (isTipoId) {
  return getTipoId(path);
}
if (isCpsId) {
  return `/${service}/live/${getCpsId(path)}`;
}

```

**Review by eagerterrier — DISMISSED** [review]
some minor changes here.

File: src/app/routes/utils/constructPageFetchUrl/index.ts
Comment (pvaliani) [reply src/app/routes/utils/constructPageFetchUrl/index.ts]: Yep, so think it's:

```
    case LIVE_PAGE:
      getIdFunction = (path: string) => {
        if (isTipoIdCheck(path)) {
          return getTipoId(path);
        }
        if (isCpsIdCheck(path)) {
          return `/${service}/live/${getCpsId(path)}`;
        }
        return null;
      };
      break;
```

File: ws-nextjs-app/integration/pages/live/arabic/canonical.test.ts
Comment (karinathomasbbc) [inline ws-nextjs-app/integration/pages/live/arabic/canonical.test.ts:1]: Can we also add an integration test for a CPS live page with variant?

File: ws-nextjs-app/integration/pages/live/arabic/amp.test.ts
Comment (pvaliani) [reply ws-nextjs-app/integration/pages/live/arabic/amp.test.ts]: Can the Pidgin amp tests file in that same location also be deleted then?

File: src/app/routes/utils/constructPageFetchUrl/index.test.ts
Comment (karinathomasbbc) [reply src/app/routes/utils/constructPageFetchUrl/index.test.ts:106]: Could we add a couple of additional scenarios for variants? Ideally for both types of Live Page (Tipo & CPS). We might need to check with the Content Team about what the expected URL of a Tipo Live page would be (i.e. is the variant at the end or in the middle of the URL?).

**Review by karinathomasbbc — CHANGES_REQUESTED** [review]

File: ws-nextjs-app/integration/pages/live/arabic/amp.test.ts
Comment (karinathomasbbc) [reply ws-nextjs-app/integration/pages/live/arabic/amp.test.ts]: Yes, we did a bit of digging in the code, and we found that the NextJS app does not currently support AMP pages. I would want some 👀 from the Content Team on this PR before we merge this, just to sense check our assumptions.

File: src/app/routes/utils/constructPageFetchUrl/index.ts
Comment (LilyL0u) [reply src/app/routes/utils/constructPageFetchUrl/index.ts:99]: Done

File: src/app/routes/utils/constructPageFetchUrl/index.test.ts
Comment (LilyL0u) [reply src/app/routes/utils/constructPageFetchUrl/index.test.ts]: Integration tests added.

File: ws-nextjs-app/integration/pages/live/arabic/canonical.test.ts
Comment (LilyL0u) [reply ws-nextjs-app/integration/pages/live/arabic/canonical.test.ts:1]: I couldn't find a variant cps live page yet. Anyone got one?

Comment (karinathomasbbc) [reply ws-nextjs-app/integration/pages/live/arabic/canonical.test.ts:1]: https://www.test.bbc.com/serbian/lat/live/media-23179005?morph_env=test 
https://www.test.bbc.com/serbian/cyr/live/media-23179005?morph_env=test

Comment (pvaliani) [reply ws-nextjs-app/integration/pages/live/arabic/canonical.test.ts:1]: Looking at the Simorgh integration tests for `serbian` `frontPage` as an example, they only seem to be for lat - how do we want them laid out in the next app for Live pages? sep folders for lat and cyr? or is it enough to just pick one variant just to ensure we are handling routes with variants? Seems like the `frontPage` was laid out that way for inty tests

Comment (karinathomasbbc) [reply ws-nextjs-app/integration/pages/live/arabic/canonical.test.ts:1]: One variant is probably enough and whichever pattern makes sense?

File: src/app/routes/utils/constructPageFetchUrl/index.ts
Comment (karinathomasbbc) [inline src/app/routes/utils/constructPageFetchUrl/index.ts]: Let's move this comment to above line 230, and then we can remove the LIVE_PAGE case wording
```suggestion
      // pathname is the ID of the Live page without /service/live/, and supports both Tipo & CPS IDs
```

File: data/serbian/live/lat/media-23179005.json
Comment (karinathomasbbc) [inline data/serbian/live/lat/media-23179005.json:1]: Remember that this needs to be renamed to data/serbian/live/media-23179005/lat.json

File: src/app/routes/utils/constructPageFetchUrl/index.ts
Comment (karinathomasbbc) [inline src/app/routes/utils/constructPageFetchUrl/index.ts]: Can this comment be moved to above line 229 i.e. directly above the code which uses `pathname` (folk might miss this comment if it's outside the `case` statement

**Review by karinathomasbbc — APPROVED** [review]

File: ws-nextjs-app/integration/pages/live/serbian/canonical.test.ts
Comment (karinathomasbbc) [inline ws-nextjs-app/integration/pages/live/serbian/canonical.test.ts]: Can we add another test in here to assert that we can see the live page title (which we could derive from the page data? I am concerned that these integration tests are "passing", when in reality, I think a 404 page is being rendered. (The data is in the wrong place, so there would be a 404 data error).

I am not sure if it exists, but I thought there was logic to ensure that any page that is under integration test returns a 200 status (except for the error pages), but maybe that's only the cypress tests...?

Comment (karinathomasbbc) [reply ws-nextjs-app/integration/pages/live/serbian/canonical.test.ts]: I think I've found a better set of tests to run: `runCommonCrossPlatformTests` instead of just the header tests (because this is only checking the header / nav bar, which will "pass" for error pages, rather than the content of the page). Can we do that before updating the serbian lat json file, and ensure that they fail?

This will run the following tests: https://github.com/bbc/simorgh/blob/596803f0a0baa3060783d328ea731ed7016321c3/src/integration/common/index.js#L32-L38

File: src/app/routes/utils/constructPageFetchUrl/index.test.ts
Comment (amoore108) [inline src/app/routes/utils/constructPageFetchUrl/index.test.ts:102]: Could add a check for variant languages here for TIPO/CPS assets?

File: src/app/components/Header/ScriptLink/index.tsx
Comment (amoore108) [inline src/app/components/Header/ScriptLink/index.tsx]: Think this is maybe a merge conflict from yesterday, resulting in it being moved above the comment below.

File: ws-nextjs-app/integration/pages/live/serbian/canonical.test.ts
Comment (pvaliani) [reply ws-nextjs-app/integration/pages/live/serbian/canonical.test.ts]: invoking `runCommonCrossPlatformTests` results in snapshot failures (as expected) when the data folder is routed with the `{id}.json` instead of `lat.json` so will update the file path now and re-run the snapshots to update:

image

File: src/app/routes/utils/constructPageFetchUrl/index.test.ts
Comment (louisearchibald) [reply src/app/routes/utils/constructPageFetchUrl/index.test.ts:102]: @amoore108 We have found a CPS live page **test** url example which I have added in a test case for, but struggling to find other variant example assets. Do you have any to hand, or able to direct me to somewhere I will be able to find some more? 😄 

(have also managed to source a **live** cps example since posting this comment)

Comment (amoore108) [reply src/app/routes/utils/constructPageFetchUrl/index.test.ts:102]: There won't be any real Test/Live TIPO assets unfortunately, but you should be able to just use something like `c0000000000t` in the other examples to mock what a TIPO asset would look like.

Comment (louisearchibald) [reply src/app/routes/utils/constructPageFetchUrl/index.test.ts:102]: Ah of course, thanks 🙂

**Comment by pvaliani** [comment]
@amoore108 can we pls have green tick blessing if you're happy with the changes? we've updated the integration tests now

**Review by amoore108 — APPROVED** [review]

**Comment by pvaliani** [comment]
@amoore108 further request but can you please discard or remove @eagerterrier change request as we've addressed those and responded but believe he's on AL until next week. Thank you!

**Pull request merged**
